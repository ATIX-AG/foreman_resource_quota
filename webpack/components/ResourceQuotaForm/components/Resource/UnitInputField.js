import React, { useState, useEffect, useCallback } from 'react';
import PropTypes from 'prop-types';
import {
  FormGroup,
  FormHelperText,
  TextInput,
  InputGroup,
  InputGroupText,
  Dropdown,
  DropdownItem,
  DropdownToggle,
} from '@patternfly/react-core';

import { sprintf, translate as __ } from 'foremanReact/common/I18n';

import { findLargestFittingUnit } from '../../../../helper';

const UnitInputField = ({
  initialValue,
  onChange,
  isDisabled,
  handleInputValidation,
  minValue,
  maxValue,
  units,
  labelIcon,
}) => {
  /* flexible unit adaption variables */
  const [errorText, setErrorText] = useState('');
  const [validated, setValidated] = useState('default');
  const [isUnitOpen, setIsUnitOpen] = useState(false);
  const bestFitUnit = findLargestFittingUnit(initialValue, units);
  const [selectedUnit, setSelectedUnit] = useState(bestFitUnit);
  const [inputValue, setInputValue] = useState(
    initialValue / bestFitUnit.factor
  );
  let unitDropdownItems = [];

  /* generate unitDropdownItems depending on unit */
  if (units.length > 1) {
    unitDropdownItems = units.map(unit => (
      <DropdownItem
        id={`unit-dropdownitem-${unit.symbol.toLowerCase()}`}
        key={unit.symbol.toLowerCase()}
      >
        {unit.symbol}
      </DropdownItem>
    ));
  }

  /* text for bounds errors */
  const errorTextBounds = useCallback(() => {
    const boundsText = 'Value must be between %d and %d.';
    let errorMin = minValue;
    let errorMax = maxValue;
    if (selectedUnit) {
      errorMin = minValue / selectedUnit.factor;
      errorMax = maxValue / selectedUnit.factor;
    }
    return __(sprintf(boundsText, errorMin, errorMax));
  }, [minValue, maxValue, selectedUnit]);

  /* text for float errors */
  const errorTextNatural = useCallback(() => __('Value must be a number.'), []);

  /* text for float inputs (rounding) */
  const warningTextRounded = useCallback(
    roundedValue => __(`Rounding to: ${roundedValue} (${units[0].symbol}).`),
    [units]
  );

  /* warning text displayed beneath value input field (built-in is used for errors) */
  const helperTextWarning = (text, isHidden) => (
    <FormHelperText isHidden={isHidden}>{text}</FormHelperText>
  );

  /* applies the selected unit and checks the bounds */
  const isValid = useCallback(
    val => {
      if (Number.isNaN(Number(val))) {
        setErrorText(errorTextNatural());
        return false;
      }
      const baseValue = valueToBaseUnit(val);
      if (baseValue < minValue || baseValue > maxValue) {
        setErrorText(errorTextBounds());
        return false;
      }
      return true;
    },
    [minValue, maxValue, valueToBaseUnit, errorTextNatural, errorTextBounds]
  );

  /* applies the selected unit and returns the base-unit value */
  const valueToBaseUnit = useCallback(
    val => {
      if (units.length > 1) {
        return selectedUnit.factor * val;
      }
      return Number(val);
    },
    [units, selectedUnit]
  );

  useEffect(() => {
    if (isDisabled) {
      handleInputValidation(true);
      setValidated('default');
    } else if (isValid(inputValue)) {
      const baseValue = valueToBaseUnit(inputValue);
      let validatedValue = baseValue;
      if (baseValue !== Math.floor(baseValue)) {
        validatedValue = Math.floor(baseValue);
        setErrorText(warningTextRounded(validatedValue));
        setValidated('warning');
      } else {
        // Keep baseValue as validatedValue
        setValidated('default');
      }
      onChange(validatedValue);
      handleInputValidation(true);
    } else {
      handleInputValidation(false);
      setValidated('error');
    }
  }, [
    isDisabled,
    inputValue,
    selectedUnit,
    handleInputValidation,
    onChange,
    isValid,
    valueToBaseUnit,
    warningTextRounded,
  ]);

  /* set the selected unit */
  const onUnitSelect = event => {
    const { id } = event.currentTarget;
    const selectedListItem = unitDropdownItems.find(
      item => item.props.id === id
    );
    const unitItem = units.find(
      item => item.symbol === selectedListItem.props.children
    );
    setSelectedUnit(unitItem);
    setIsUnitOpen(false);
    // FIXME: Fix input validation on unit selection
  };

  const onUnitToggle = () => {
    setIsUnitOpen(!isUnitOpen);
  };

  /* return the unit view depending on the available units
   *
   * unit = null : don't print any unit
   * unit = [.]  : print a single Textfield
   * unit = [..] : print an editable Dropdown menu
   * */
  const unitView = () => {
    if (units.length === 1) {
      return <InputGroupText>{__(units[0].symbol)}</InputGroupText>;
    }

    return (
      <Dropdown
        onSelect={onUnitSelect}
        toggle={
          <DropdownToggle isDisabled={isDisabled} onToggle={onUnitToggle}>
            {__(`${selectedUnit.symbol}`)}
          </DropdownToggle>
        }
        isOpen={isUnitOpen}
        dropdownItems={unitDropdownItems}
      />
    );
  };

  return (
    <FormGroup
      label={__('Quota Limit')}
      validated={validated}
      helperTextInvalid={errorText}
      helperText={helperTextWarning(errorText, validated !== 'warning')}
      fieldId="quota-limit-resource-quota-form-group"
      labelIcon={labelIcon || {}}
    >
      <InputGroup>
        <TextInput
          isDisabled={isDisabled}
          value={inputValue}
          min={minValue}
          max={maxValue}
          validated={validated}
          id="reg_token_life_time_input"
          onChange={setInputValue}
        />
        {unitView()}
      </InputGroup>
    </FormGroup>
  );
};

UnitInputField.propTypes = {
  initialValue: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  onChange: PropTypes.func.isRequired,
  isDisabled: PropTypes.bool.isRequired,
  handleInputValidation: PropTypes.func.isRequired,
  units: PropTypes.arrayOf(
    PropTypes.shape({
      symbol: PropTypes.string,
      factor: PropTypes.number,
    })
  ).isRequired,
  labelIcon: PropTypes.node,
  minValue: PropTypes.number.isRequired,
  maxValue: PropTypes.number.isRequired,
};

UnitInputField.defaultProps = {
  initialValue: 0,
  labelIcon: null,
};

export default UnitInputField;
