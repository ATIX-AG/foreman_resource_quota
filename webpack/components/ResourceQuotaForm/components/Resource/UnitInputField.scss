@import 'foremanReact/common/variables';

.container-unit-input-field {
  .pf-v5-c-input-group__text {
    word-break: keep-all;
  }
}
