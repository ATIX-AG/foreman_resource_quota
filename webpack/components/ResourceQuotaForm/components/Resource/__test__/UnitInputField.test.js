import React from 'react';
import '@testing-library/jest-dom';
import { render, screen, fireEvent } from '@testing-library/react';

import { testComponentSnapshotsWithFixtures } from '@theforeman/test';
import LabelIcon from 'foremanReact/components/common/LabelIcon';

import UnitInputField from '../UnitInputField';

const getDefaultProps = () => ({
  initialValue: 0,
  onChange: jest.fn(),
  isDisabled: false,
  handleInputValidation: jest.fn(),
  units: [
    { symbol: 'MiB', factor: 1 },
    { symbol: 'GiB', factor: 1024 },
  ],
  labelIcon: <LabelIcon text="Descriptive title." />,
  minValue: 0,
  maxValue: 5,
});

const fixtureDefault = {
  'should render default': {
    ...getDefaultProps(),
  },
};

const fixtureSingleUnit = {
  'should render without dropdown (single unit)': {
    ...getDefaultProps(),
    units: [{ symbol: 'cores', factor: 1 }],
  },
};

const fixtureDisabled = {
  'should render as disabled field': {
    ...getDefaultProps(),
    isDisabled: true,
  },
};

describe('UnitInputField', () => {
  testComponentSnapshotsWithFixtures(UnitInputField, fixtureDefault);
  testComponentSnapshotsWithFixtures(UnitInputField, fixtureSingleUnit);
  testComponentSnapshotsWithFixtures(UnitInputField, fixtureDisabled);

  it('triggers handleInputValidation on unit change', async () => {
    const props = getDefaultProps();

    render(<UnitInputField {...props} />);
    const input = screen.getByRole('textbox');
    fireEvent.change(input, { target: { value: 3 } });

    // gets called (1.) with initialValue and (2.) the simulated change
    expect(props.onChange).toHaveBeenCalledTimes(2);
    expect(props.onChange).toHaveBeenCalledWith(props.initialValue);
    expect(props.onChange).toHaveBeenLastCalledWith(3);
    expect(props.handleInputValidation).toHaveBeenCalledTimes(2);
    expect(props.handleInputValidation).toHaveBeenCalledWith(true);
  });

  test('triggers onChange with rounded value', () => {
    const props = getDefaultProps();

    render(<UnitInputField {...props} />);
    const input = screen.getByRole('textbox');
    fireEvent.change(input, { target: { value: 3.5 } });

    // gets called (1.) with initialValue and (2.) the simulated change
    expect(props.onChange).toHaveBeenCalledTimes(2);
    expect(props.onChange).toHaveBeenCalledWith(props.initialValue);
    expect(props.onChange).toHaveBeenLastCalledWith(3);
    expect(props.handleInputValidation).toHaveBeenCalledTimes(2);
    expect(props.handleInputValidation).toHaveBeenCalledWith(true);
  });

  test('does not trigger onChange when value out of bounds', () => {
    const props = getDefaultProps();

    render(<UnitInputField {...props} />);
    const input = screen.getByRole('textbox');
    fireEvent.change(input, { target: { value: props.maxValue + 1 } });

    // onChange only called for initialValue
    expect(props.onChange).toHaveBeenCalledTimes(1);
    expect(props.onChange).toHaveBeenCalledWith(props.initialValue);
    // handleInputValidation called with false => invalid
    expect(props.handleInputValidation).toHaveBeenCalledTimes(2);
    expect(props.handleInputValidation).toHaveBeenLastCalledWith(false);
  });

  test('does not trigger onChange when value is not a number', () => {
    const props = getDefaultProps();

    render(<UnitInputField {...props} />);
    const input = screen.getByRole('textbox');
    fireEvent.change(input, { target: { value: 'no number' } });

    // onChange only called for initialValue
    expect(props.onChange).toHaveBeenCalledTimes(1);
    expect(props.onChange).toHaveBeenCalledWith(props.initialValue);
    // handleInputValidation called with false => invalid
    expect(props.handleInputValidation).toHaveBeenCalledTimes(2);
    expect(props.handleInputValidation).toHaveBeenLastCalledWith(false);
  });
});
