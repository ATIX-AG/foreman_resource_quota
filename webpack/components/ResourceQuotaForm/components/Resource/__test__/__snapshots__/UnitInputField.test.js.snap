// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UnitInputField should render as disabled field 1`] = `
<FormGroup
  fieldId="quota-limit-resource-quota-form-group"
  helperText={
    <FormHelperText
      isHidden={true}
    >
      
    </FormHelperText>
  }
  helperTextInvalid=""
  label="Quota Limit"
  labelIcon={
    <LabelIcon
      text="Descriptive title."
    />
  }
  validated="default"
>
  <InputGroup>
    <TextInput
      id="reg_token_life_time_input"
      isDisabled={true}
      max={5}
      min={0}
      onChange={[Function]}
      validated="default"
      value={0}
    />
    <Dropdown
      dropdownItems={
        Array [
          <DropdownItem
            id="unit-dropdownitem-mib"
          >
            MiB
          </DropdownItem>,
          <DropdownItem
            id="unit-dropdownitem-gib"
          >
            GiB
          </DropdownItem>,
        ]
      }
      isOpen={false}
      onSelect={[Function]}
      toggle={
        <DropdownToggle
          isDisabled={true}
          onToggle={[Function]}
        >
          MiB
        </DropdownToggle>
      }
    />
  </InputGroup>
</FormGroup>
`;

exports[`UnitInputField should render default 1`] = `
<FormGroup
  fieldId="quota-limit-resource-quota-form-group"
  helperText={
    <FormHelperText
      isHidden={true}
    >
      
    </FormHelperText>
  }
  helperTextInvalid=""
  label="Quota Limit"
  labelIcon={
    <LabelIcon
      text="Descriptive title."
    />
  }
  validated="default"
>
  <InputGroup>
    <TextInput
      id="reg_token_life_time_input"
      isDisabled={false}
      max={5}
      min={0}
      onChange={[Function]}
      validated="default"
      value={0}
    />
    <Dropdown
      dropdownItems={
        Array [
          <DropdownItem
            id="unit-dropdownitem-mib"
          >
            MiB
          </DropdownItem>,
          <DropdownItem
            id="unit-dropdownitem-gib"
          >
            GiB
          </DropdownItem>,
        ]
      }
      isOpen={false}
      onSelect={[Function]}
      toggle={
        <DropdownToggle
          isDisabled={false}
          onToggle={[Function]}
        >
          MiB
        </DropdownToggle>
      }
    />
  </InputGroup>
</FormGroup>
`;

exports[`UnitInputField should render without dropdown (single unit) 1`] = `
<FormGroup
  fieldId="quota-limit-resource-quota-form-group"
  helperText={
    <FormHelperText
      isHidden={true}
    >
      
    </FormHelperText>
  }
  helperTextInvalid=""
  label="Quota Limit"
  labelIcon={
    <LabelIcon
      text="Descriptive title."
    />
  }
  validated="default"
>
  <InputGroup>
    <TextInput
      id="reg_token_life_time_input"
      isDisabled={false}
      max={5}
      min={0}
      onChange={[Function]}
      validated="default"
      value={0}
    />
    <InputGroupText>
      cores
    </InputGroupText>
  </InputGroup>
</FormGroup>
`;
